LUN-208: Feedback for a word

f1sh1918 opened the pull request:
This pull request belongs to an issue on our [bugtracker](https://issues.tuerantuer.org/secure/RapidBoard.jspa?rapidView=27&view=detail).
You can find it there by looking for an issue with the key which is mentioned in the title of this pull request.
It starts with the keyword _LUN_.

Remove comment mentioned in the code (ExerciseHeader.tsx LN:57) to test modal
1. go to an exercise
2. click on kebab menu
3. open feedback
4. type message
5. send or close modal (submit is not implemented yet since backend is missing)

Also check the refactored `CustomTextInput` in `WriteExercise`, `AddCustomDiscipline`

f1sh1918 commented on src/components/ExerciseHeader.tsx:
```diff
@@ -37,9 +51,17 @@ const ExerciseHeader = ({ navigation, currentWord, numberOfWords }: ExerciseHead
+        headerRight: () => (
+          <HeaderRightContainer>
+            <StyledContent>{progressText}</StyledContent>
+            {/* TODO Remove comment when LUNES-269 is ready */}
```
remove this comment for testing

f1sh1918 commented on src/components/HeaderWithMenu.tsx:
```diff
@@ -48,19 +48,17 @@ const HeaderWithMenu = ({ navigation }: Props): JSX.Element => (
-            />
-          </OverflowMenu>
-        </HeaderButtons>
+        <KebabMenu
```
Since kebab menu is now used in several place, i created a separate component

SteffiStuffel commented on src/components/KebabMenu.tsx:
```diff
@@ -0,0 +1,19 @@
+
+interface KebabMenuProps {
+  children: ReactElement
+  icon: ReactElement
```
(optional) As I assume our kebab menu always has the three icons, we maybe can just add an optional boolean param like whiteIcon? or something?

f1sh1918 replied:
i found out a better thing, you can just use "currentColor" and use the color attribute for styling, so we only need one icon not white and primary
```js
<svg width="7" height="29" viewBox="0 0 7 29" xmlns="http://www.w3.org/2000/svg">
    <g fill="currentColor">
            <circle cx="3.5" cy="3.5" r="3.5"/>
            <circle cx="3.5" cy="14.5" r="3.5"/>
            <circle cx="3.5" cy="25.5" r="3.5"/>
    </g>
</svg>

f1sh1918 replied:
in general yes you could also use `fill="currentColor"` in the svg to style in with the css attrib `color: primary` f.e. but we use different sizes and padding f.e. in HeaderWithMenu, so i would like to keep it as an ReactElement but refactor the svg to avoid several svg only differing in color

SteffiStuffel commented on src/components/__tests__/FeedbackModal.spec.tsx:
```diff
@@ -0,0 +1,40 @@
+describe('FeedbackModal', () => {
+  const setVisible = jest.fn()
+
+  it('Feedback should only be conveyable if a message was set', () => {
```
(minor) maybe just "Button should be disabled for empty message"

SteffiStuffel commented on src/components/__tests__/FeedbackModal.spec.tsx:
```diff
@@ -0,0 +1,40 @@
+    fireEvent.press(submitButton)
+    expect(setVisible).toHaveBeenCalledWith(false)
+  })
+  it('Clears feedback text by clicking clear button', () => {
```
(minor) new lines between test

SteffiStuffel commented on src/constants/labels.json:
```diff
@@ -109,13 +109,21 @@
+  "feedback": {
+    "sendFeedback": "Senden",
+    "question": "Hast du Feedback oder einen Wortvorschlag?",
+    "mailPlaceholder": "Deine Email für Fragen(optional)",
```
space between Fragen and (optional) would be nice, although it is missing in the designs.

SteffiStuffel commented on src/components/CustomTextInput.tsx:
```diff
@@ -0,0 +1,123 @@
+          onSubmitEditing={onSubmitEditing}
+          textAlignVertical={multiLine ? 'top' : 'center'}
+        />
+        <IconContainer multiLine={multiLine}>
```
When adding a custom discipline there previously was no close-icon to delete the input, because at the same position is the scan-qr-code-icon. Please change, so the scan-qr-code-icon is always visible

f1sh1918 replied:
alright i thought maybe it would be nice to have a remove option after scanning/typing key. Because it's way more handy on mobile devices to have a single click remove option than to focus input field and remove with software keyboard. so for scanning it would just one click to remove content and then click on the scan symbol

SteffiStuffel replied:
I think we should stick to the old behaviour, as a user cannot know that he needs to delete the input to be able to scan again.

f1sh1918 replied:
i removed the `clearable`

SteffiStuffel commented on src/routes/add-custom-discipline/AddCustomDisciplineScreen.tsx:
```diff
@@ -111,21 +89,20 @@ const AddCustomDiscipline = ({ navigation }: AddCustomDisciplineScreenProps): JS
+              clearable
+              rightContainer={
+                <TouchableOpacity onPress={() => setShowQRCodeOverlay(true)}>
+                  <QRCodeIcon accessibilityLabel='qr-code-scanner' width={wp('6%')} height={wp('6%')} />
```
styling of the qr-code icon is now wrong. The padding on the right is too small.

SteffiStuffel reviewed (commented):
Nice work :tada: 

two things I recognized:
1) It should be possible to give feedback also in the word list, both in the overview of the words and also in the detail view
2) adding qr-codes is now broken, I always get error messages also for valid api-keys

SteffiStuffel commented:
An the focused/non-focues borders are almost identically. Is this intended? e.g. in Write exercise I could not see the color changes.

f1sh1918 commented on jest.config.ts:
```diff
@@ -1,4 +1,4 @@
-const transformNodeModules = ['react-native', '@react-native']
+const transformNodeModules = ['react-native', '@react-native', 'react-navigation-header-buttons']
```
because the jests fail when you import this library because a module (of this lib) can't be exported if i remember right. i can provide some error message if you want.

SteffiStuffel replied:
not it's fine, thanks.

f1sh1918 commented on src/components/CustomTextInput.tsx:
```diff
@@ -0,0 +1,123 @@
+          )}
+        </IconContainer>
+      </TextInputContainer>
+      {showErrorValidation && (
```
hm i think a text input base component should have some kind of field validation (and error message) because we may need it in the future when the user sends more info to the cms. I can refactor it to be more generic for Write-Exercise, i just didn't want to make it too complicated

f1sh1918 commented:
> An the focused/non-focues borders are almost identically. Is this intended? e.g. in Write exercise I could not see the color changes.

hm i'm testing the version before my refactorting and couldn't see any focus border changes onFocus before

f1sh1918 commented on src/components/CustomTextInput.tsx:
```diff
@@ -0,0 +1,123 @@
+
+// depending on error and focus the border color will be defined
+const getBorderColor = (hasErrorMessage: boolean, isFocused: boolean): string => {
+  if (hasErrorMessage) {
```
you mean to pass f.e. for `AddCustomDisciplineScreen`
```js
customBorderColor={errorMessage.length> 0 ? theme.colors.incorrect: undefined}
```
well on the one hand you are right it could be confusing but on the other hand to pass every time a border color for this standard error validation is kind of a pain. it should be included in the errorValidation handled by the component itself in my opition

f1sh1918 replied:
since there we were not more comments and we said in case of two different opinions the creator the pr wins, i mark it as resolved.

f1sh1918 commented on src/components/ConfirmationModal.tsx:
```diff
@@ -39,27 +39,47 @@ const Message = styled(HeadingText)`
 }
 
+// TODO Further adjustments gonna be done with LUN-312
 const ConfirmationModal = (props: ConfirmationModalProps): JSX.Element => {
```
gonna check this and resolve the conflicts

f1sh1918 commented:
Updated branch.
Added Kebab Menu and Feedback Modal for VocabularyList and VocDetail
Search for comments with `LUNES-269` and uncomment the parts.

Due to some conflicts with UnlockExercise Component, i switched the buttons for the confirmation modal to be consistent (1. Button = confirmation, 2nd Button = cancel)

Some additional label changes for the button can be done with https://issues.tuerantuer.org/browse/LUN-312.
If thats fine.

SteffiStuffel commented:
Hidden comments are not yet resolved.

Please use correct issue reference LUN-xxx is for development tasks. LUNES-xx are the orga board issues

f1sh1918 commented on src/components/FeedbackModal.tsx:
```diff
@@ -0,0 +1,61 @@
+    setVisible(false)
+  }
+  const onSubmit = (): void => {
+    // TODO Submit Feedback LUNES-269
```
sry haven't seen it yet

f1sh1918 commented on src/components/FeedbackModal.tsx:
```diff
@@ -0,0 +1,61 @@
+
+interface FeedbackModalProps {
+  visible: boolean
+  setVisible: (visible: boolean) => void
```
> Hidden comments are not yet resolved.
> 
> Please use correct issue reference LUN-xxx is for development tasks. LUNES-xx are the orga board issues

sorry i really overseen your comments. hope now its fine :)

SteffiStuffel commented on src/components/ConfirmationModal.tsx:
```diff
@@ -39,56 +39,48 @@ const Message = styled(HeadingText)`
 
 export interface ConfirmationModalProps {
```
If the purpose of this component is not longer to show a confirmation-question which can be answered with "oke" or "cancel", the naming of this component should no longer be `ConfirmationModal`

f1sh1918 replied:
ok we can use CustomModal

SteffiStuffel commented on src/components/CustomTextInput.tsx:
```diff
@@ -0,0 +1,123 @@
+const LINE_HEIGHT = 30
+const MIN_HEIGHT = 60
+
+const StyledTextInput = styled.TextInput`
```
Please use the lunes text components. You can use `ContentTextLight` here and remove font-size, font-weight and font-family then.

f1sh1918 replied:
this is a styled `textInput` not `text`

SteffiStuffel replied:
sorry, you are right. Just fine as it is.

SteffiStuffel commented on src/components/CustomTextInput.tsx:
```diff
@@ -0,0 +1,123 @@
+const TextInputContainer = styled.View<{ lines: number; borderColor: string; showErrorValidation: boolean }>`
+  border: 1px solid ${props => props.borderColor};
+  padding-left: ${props => props.theme.spacings.sm};
+  padding-right: ${props => props.theme.spacings.sm};
```
(optional) can be simplified to 
`padding: 0 ${props => props.theme.spacings.sm}`

SteffiStuffel commented on src/components/CustomTextInput.tsx:
```diff
@@ -0,0 +1,123 @@
+interface CustomTextInputProps extends TextInputProps {
+  value: string
+  onChangeText: (value: string) => void
+  /** calculates the height for textArea */
```
please remove this comments. If naming does not make clear what the prop does, rename it.

SteffiStuffel commented on src/components/CustomTextInput.tsx:
```diff
@@ -0,0 +1,123 @@
+  customBorderColor?: string
+}
+
+// depending on error and focus the border color will be defined
```
please remove this comment, what the function does it quite obvious.

SteffiStuffel commented on src/components/CustomTextInput.tsx:
```diff
@@ -0,0 +1,123 @@
+  return isFocused ? theme.colors.primary : theme.colors.textSecondary
+}
+
+const CustomTextInput: React.FC<CustomTextInputProps> = ({
```
Why this `React.FC<..:>` type? We do not use this anywhere else, and it has no additional information. Please remove.

f1sh1918 replied:
your right this is an old standard. have to adjust my snippets :)

SteffiStuffel commented on src/components/__tests__/FeedbackModal.spec.tsx:
```diff
@@ -0,0 +1,42 @@
+describe('FeedbackModal', () => {
+  const onClose = jest.fn()
+
+  it('Button should be disabled for empty message', () => {
```
test naming always starts with "should", please rename tests.

steffenkleinle replied:
should also be lowercase

f1sh1918 replied:
done

SteffiStuffel commented on src/constants/labels.json:
```diff
@@ -17,7 +17,7 @@
     "deleteModal": {
       "confirmationQuestion": "Soll der Bereich wirklich gelöscht werden?",
-      "cancel": "Abbrechen",
+      "cancel": "Schließen",
```
now we have 3 times the "cancel": "Schließen", please only keep it once and move it to "general"

f1sh1918 replied:
removed the three entries and created one in general

SteffiStuffel reviewed (changes requested):
For me it is not working to open the Menu in VocabularyList detail view.

https://user-images.githubusercontent.com/18463686/169249718-9e361f9d-b8c7-4790-9dee-74736bc4347c.mp4

f1sh1918 commented on src/components/ConfirmationModal.tsx:
```diff
@@ -39,56 +39,48 @@ const Message = styled(HeadingText)`
-
   confirmationButtonText: string
-  cancelButtonText: string
+  cancelButtonText?: string
```
still waiting for feedback if it really should be the same string

SteffiStuffel replied:
yes please always use "Zurück".

f1sh1918 commented:
> For me it is not working to open the Menu in VocabularyList detail view.
> 
>  LUN-208-feedback.mp4

hm seems like overflowMenu over a modal works only on ios. tried a lot of styling but didn't find a way yet

steffenkleinle commented on src/components/CustomTextInput.tsx:
```diff
@@ -0,0 +1,118 @@
+}: CustomTextInputProps): ReactElement => {
+  const theme = useTheme()
+  const [isFocused, setIsFocused] = useState<boolean>(false)
+  const hasErrorMessage = !!(errorMessage && errorMessage.length > 0)
+  const showErrorValidation = errorMessage !== undefined
```
Do we really need both? I think this is quite similar and I think we don't have to handle `errorMessage.length === 0` specifically.

f1sh1918 replied:
yes we have because we render the container as a placeholder if there is no error message to reserve the height for an error message

SteffiStuffel replied:
I also think this should be adjusted here.
1. Why are you testing for undefined one time with `!!errormessage` and one time with `errorMessage !== undefined`
2. Why at all introducing the hasErrorMessage variable which is only used once

Renaming of `showErrorValidation` would also make sense, so it is obvious what the variable does. 

I think it is not really a good design, that an empty string passing has a meaning on the design.

f1sh1918 replied:
1. we can use in both cases `errorMessage !== undefined` if i just use `errorMessage` its false for an empty string and the ErrorContainer (Placeholder) will not be displayed
2.  removed var

I understand the opinion that an empty string shouldn't change the design. So what you suppose, a separate boolean var like `hasErrorValidation` or sth similar? And then pass two separate props?

steffenkleinle commented on src/components/CustomTextInput.tsx:
```diff
@@ -0,0 +1,118 @@
+              <CloseIcon width={wp('6%')} height={wp('6%')} />
+            </ClearContainer>
+          ) : (
+            rightContainer
```
Not sure if `rightCONTAINER` is good naming, why is it relevant that it is/is not a container?

f1sh1918 replied:
its not relevant but what are your suggestion for naming?

steffenkleinle replied:
just rightContent or similar?

steffenkleinle commented on src/components/CustomTextInput.tsx:
```diff
@@ -0,0 +1,118 @@
+        </IconContainer>
+      </TextInputContainer>
+      {showErrorValidation && (
+        <ErrorContainer>{errorMessage.length > 0 && <ContentError>{errorMessage}</ContentError>}</ErrorContainer>
```
If you just use `hasErrorMessage` in the line above you don't have to check for `errorMessage.length > 0`

f1sh1918 replied:
that doesn't work because if you want to `showErrorValidation` you use the ErrorContainer as a placeholder even if there is currently no errorMessage. So you don't have jumping elements

steffenkleinle commented on src/components/HeaderWithMenu.tsx:
```diff
@@ -1,11 +1,12 @@
 import { StackNavigationProp } from '@react-navigation/stack'
```
I think `HeaderWithMenu` is not good naming anymore, this is specific for the home screen, so imo `HomeScreenHeader` would be better.

steffenkleinle commented on src/components/KebabMenu.tsx:
```diff
@@ -0,0 +1,19 @@
+import React, { ReactElement } from 'react'
```
Is this really officially called `KebabMenu`? In material design this is called `OverflowMenu`: https://material.io/components/app-bars-top#anatomy

f1sh1918 replied:
i can rename it the problem is that we have `OverflowMenu` twice then

SteffiStuffel replied:
I would also prefer OverflowMenu

steffenkleinle replied:
just import it as something else, or where do we have it as well?

f1sh1918 replied:
In the `KebabMenu` i import the RN-OverflowMenu
adjusted the import
```import { HeaderButtons, OverflowMenu as OverflowMenuRN } from 'react-navigation-header-buttons'```

steffenkleinle commented:
Personally I think I am more used to the confirmation action button to be at the bottom. I am always clicking on cancel, but maybe its just me, but it would correspond to material design to change the order:
https://material.io/components/dialogs#anatomy
While it is horizontally aligned and we have vertically aligned buttons it should apply anyway.

SteffiStuffel commented:
> Personally I think I am more used to the confirmation action button to be at the bottom. I am always clicking on cancel, but maybe its just me, but it would correspond to material design to change the order: https://material.io/components/dialogs#anatomy While it is horizontally aligned and we have vertically aligned buttons it should apply anyway.

We decided to have the confirm action at the top, as it is suggested by material.

steffenkleinle commented:
> > Personally I think I am more used to the confirmation action button to be at the bottom. I am always clicking on cancel, but maybe its just me, but it would correspond to material design to change the order: https://material.io/components/dialogs#anatomy While it is horizontally aligned and we have vertically aligned buttons it should apply anyway.
> 
> We decided to have the confirm action at the top, as it is suggested by material.

Do you have a link please?

SteffiStuffel commented:
> > > Personally I think I am more used to the confirmation action button to be at the bottom. I am always clicking on cancel, but maybe its just me, but it would correspond to material design to change the order: https://material.io/components/dialogs#anatomy While it is horizontally aligned and we have vertically aligned buttons it should apply anyway.
> > 
> > 
> > We decided to have the confirm action at the top, as it is suggested by material.
> 
> Do you have a link please?

[To the discussion or to material?](https://chat.tuerantuer.org/digitalfabrik/pl/3pmgkecewtdkze35u1brrhkxia)

steffenkleinle commented:
Please fix this or create an issue for it if it was already present before your PR:
![image](https://user-images.githubusercontent.com/29095487/170048720-6354905c-d1a3-483f-b1f1-17bf4a009ec7.png)

f1sh1918 commented:
> Please fix the overflow menu overlapping the status bar or create an issue for it if it was already present before your PR: 

can not reproduce issue, pls check updated branch (there were some updates from main meanwhile)

<img alt="" width="400px" src="https://user-images.githubusercontent.com/37902063/170055991-a1dc7fff-ce2c-4cf4-a12c-a5198398dc36.png">

f1sh1918 commented on src/routes/write-exercise/components/InteractionSection.tsx:
```diff
@@ -126,7 +109,7 @@ const InteractionSection = (props: InteractionSectionProps): ReactElement => {
       }
     }
-    return isFocused ? theme.colors.primary : theme.colors.textSecondary
+    return theme.colors.primary
```
yes check line 59 in CustomTextInput

f1sh1918 commented:
I hope i resolved all issues now. If there are better suggestions for errorValidation, feel free to add.
I removed the OverflowMenu for VocabularyDetailScreen and created a ticket for refactoring the Vocabulary Detail Screen
https://issues.tuerantuer.org/browse/LUN-357

steffenkleinle commented on src/components/OverflowMenu.tsx:
```diff
@@ -0,0 +1,19 @@
+
+import labels from '../constants/labels.json'
+
+interface KebabMenuProps {
```
```suggestion
interface OverflowMenuProps {
```

steffenkleinle reviewed (approved):
Can't test the header again right now, if you say this does not occur fine for me.

steffenkleinle commented on src/components/HomeScreenHeader.tsx:
```diff
@@ -1,11 +1,12 @@
 import { StackNavigationProp } from '@react-navigation/stack'
```
Because I requested it as the previous name `HeaderWithMenu` was definitely misleading. This is definitely home screen specific as the option `Manage disciplines` does not make sense somewhere else.

steffenkleinle commented:
:tada:
